File: mobile/android/core/auth/src/main/kotlin/com/slopssaloon/omen/core/auth/GoTrueTransport.kt
package com.slopssaloon.omen.core.auth

/**
 * Thin seam over the Supabase GoTrue REST endpoints (M0c §2). The concrete HTTP+JSON transport
 * ([app] `OkHttpGoTrueTransport`) lives in the app module so `core:auth` stays dependency-light
 * and JVM-unit-testable — [SupabaseAuthRepository]'s outcome mapping is verified here with a
 * fake transport, while the transport itself is proven by build + live smoke.
 *
 * The transport returns already-classified, **opaque** results: it never surfaces raw provider
 * bodies, tokens beyond the session fields, or identifiers (M0c §8).
 */
interface GoTrueTransport {
    suspend fun requestEmailOtp(email: String): TransportResult
    suspend fun verifyEmailOtp(email: String, code: String): TransportResult
    suspend fun signInWithGoogleIdToken(idToken: String, rawNonce: String): TransportResult
    suspend fun refresh(refreshToken: String): TransportResult

    // M4-Auth-Providers-v1 §3.2 — provider-agnostic OAuth code exchange.
    suspend fun exchangeOAuthCode(providerId: String, code: String, codeVerifier: String): TransportResult

    // M4-Auth-Providers-v1 §3.2 — WebAuthn challenge issue + assertion verify + registration.
    suspend fun startPasskeyChallenge(): TransportResult
    suspend fun verifyPasskeyAssertion(assertion: PasskeyResult.Assertion): TransportResult
    suspend fun registerPasskey(credential: PasskeyResult.Assertion): TransportResult
}

sealed interface TransportResult {
    /** 2xx with no session body (e.g. OTP send accepted). */
    data object Ok : TransportResult

    /**
     * 2xx returning a WebAuthn challenge (opaque base64url string). Only used by
     * `startPasskeyChallenge`; other endpoints never produce this variant.
     */
    data class Challenge(val challenge: String) : TransportResult

    /** 2xx with a session. [expiresInSeconds] is the token lifetime from `now`. */
    data class SessionTokens(
        val userId: String,
        val accessToken: String,
        val refreshToken: String,
        val expiresInSeconds: Long,
    ) : TransportResult

    /** Non-2xx HTTP status. Only the numeric [status] is carried — never the body. */
    data class HttpError(val status: Int) : TransportResult

    /** Transport-level failure (no response, socket, DNS, timeout). */
    data object NetworkError : TransportResult

    /** A 2xx body that could not be parsed into the expected shape. */
    data object Malformed : TransportResult
}

File: mobile/ios/OmenIOS/OmenIOS/Core/Auth/GoTrueTransport.swift
import Foundation

/// Mirrors Android `core/auth/GoTrueTransport.kt`'s `TransportResult`. The lowest-level view of
/// a GoTrue HTTP call, before any outcome mapping.
enum TransportResult: Equatable {
    case ok
    case sessionTokens(userID: String, accessToken: String, refreshToken: String, expiresInSeconds: Int)
    case httpError(status: Int)
    case networkError
    case malformed
}

enum PasskeyOptionsTransportResult<Options: Equatable>: Equatable {
    case options(Options)
    case httpError(status: Int)
    case networkError
    case malformed
}

enum PasskeyListTransportResult: Equatable {
    case passkeys([PasskeyInfo])
    case httpError(status: Int)
    case networkError
    case malformed
}

/// Mirrors Android `core/auth/GoTrueTransport.kt`. The production implementation is
/// `URLSessionGoTrueTransport` (App layer, needs `URLSession`); this protocol has no framework
/// dependency so it can live in Core and be faked in tests.
///
/// `signInWithIDToken` is generalized over `provider` (rather than a Google-only method) because
/// iOS's primary native mechanism is Apple, not Google — same endpoint
/// (`POST /auth/v1/token?grant_type=id_token`), different `provider` value in the body.
protocol GoTrueTransport {
    func requestEmailOtp(email: String) async -> TransportResult
    func verifyEmailOtp(email: String, code: String) async -> TransportResult
    func signInWithIDToken(provider: String, idToken: String, nonce: String?) async -> TransportResult
    func refresh(refreshToken: String) async -> TransportResult

    // M4-Auth-Providers-v1 §4.2 — provider-agnostic OAuth code exchange.
    func exchangeOAuthCode(providerId: String, code: String, codeVerifier: String) async -> TransportResult

    // Supabase Auth passkey endpoints. Registration and management require the current user's
    // access token; authentication is a discoverable, signed-out ceremony.
    func startPasskeyAuthentication() async -> PasskeyOptionsTransportResult<PasskeyAuthenticationOptions>
    func verifyPasskeyAuthentication(challengeID: String, assertion: PasskeyResult.Assertion) async -> TransportResult
    func startPasskeyRegistration(accessToken: String) async -> PasskeyOptionsTransportResult<PasskeyRegistrationOptions>
    func verifyPasskeyRegistration(challengeID: String, credential: PasskeyRegistrationResult.Credential, accessToken: String) async -> TransportResult
    func listPasskeys(accessToken: String) async -> PasskeyListTransportResult
    func deletePasskey(id: String, accessToken: String) async -> TransportResult
}
